packages/client/lib/rpc/modules/eth.ts

```diff
@@ -39,24 +73,178 @@ export class Eth {
+   * Note that the estimate may be significantly more than the amount of gas actually used by the transaction,
+   * for a variety of reasons including EVM mechanics and node performance.
+   * Currently only "latest" block is supported.
+   * @param params An array of two parameters:
+   *   1. The transaction object
+   *       * from (optional) - The address the transaction is sent from
+   *       * to - The address the transaction is directed to
+   *       * gas (optional) - Integer of the gas provided for the transaction execution
+   *       * gasPrice (optional) - Integer of the gasPrice used for each paid gas
+   *       * value (optional) - Integer of the value sent with this transaction
+   *       * data (optional) - Hash of the method signature and encoded parameters.
+   *   2. integer block number, or the string "latest", "earliest" or "pending"
+   * @returns The amount of gas used.
+   */
+  async estimateGas(params: [RpcCallTx, string]) {
+    const [transaction, blockOpt] = params
+
+    if (blockOpt !== 'latest') {
+      const latest = await this.blockNumber()
+      if (blockOpt !== latest) {
+        return {
+          code: INVALID_PARAMS,
+          message: `Currently only "latest" block supported`,
+        }
+      }
+    }
+
+    if (!this._vm) {
+      throw new Error('missing vm')
+    }
+
+    if (!transaction.gas) {
+      // If no gas limit is specified use the last block gas limit as an upper bound.
+      const latestHeader = await this._chain.getLatestHeader()
+      transaction.gas = latestHeader.gasLimit as any
+    }
+
+    const vm = this._vm.copy()
+    const txData = { ...transaction, gasLimit: transaction.gas }
+    const tx = Transaction.fromTxData(txData, { common: vm._common, freeze: false })
+
+    // set from address
+    const from = transaction.from ? Address.fromString(transaction.from) : Address.zero()
+    tx.getSenderAddress = () => {
+      return from
+    }
+
+    const { gasUsed } = await vm.runTx({
+      tx,
+      skipNonce: true,
+      skipBalance: true,
+      skipBlockGasLimitValidation: true,
+    })
+    return `0x${gasUsed.toString(16)}`
+  }
+
+  /**
+   * Returns the balance of the account at the given address.
+   * Currently only "latest" block is supported.
+   * @param params An array of two parameters:
+   *   1. address of the account
+   *   2. integer block number, or the string "latest", "earliest" or "pending"
+   */
+  async getBalance(params: [string, string]) {
+    const [addressHex, blockOpt] = params
+
+    if (blockOpt !== 'latest') {
+      const latest = await this.blockNumber()
+      if (blockOpt !== latest) {
+        return {
+          code: INVALID_PARAMS,
+          message: `Currently only "latest" block supported`,
+        }
+      }
+    }
+
+    if (!this._vm) {
+      throw new Error('missing vm')
+    }
+
+    const address = Address.fromString(addressHex)
+    const account: Account = await (this._vm.stateManager as any).getAccount(address)
```

Comment:
I think also here (and also other variants) we want to also copy VM. Between the `await this.blockNumber()` and actually getting the account, VM might execute a block, and therefore return the wrong value (it is not `latest`). What we could do is make an explicit VM copy in the constructor, and then use semaphores to ensure we are not running two queries on the copied VM twice.

Reply:
hm good point. i'm not sure about copying in the constructor because then it would only be on the block that the rpc server was created. i think we can figure out a solution here though.

Reply:
decided to copy the vm first, let me know what you think, https://github.com/ethereumjs/ethereumjs-monorepo/pull/1130/commits/d4849c0dcfe426d9bfa1a69ec2aaa1932532194e

Reply:
Copying makes sense to me as well... maybe in subsequent PR that adds support for specifying which block to execute the calls at this could be re-visited though. 

Just peeked at how ganache and hardhat handle this and it seems like
+ ganache [creates a new vm][1] 
+ hardhat [sets/unsets the StateManager state root][2] 

[1]: https://github.com/trufflesuite/ganache-core/blob/8a7ba780a664e32b2210e32c3bfebe359e7cfc37/src/chains/ethereum/ethereum/src/blockchain.ts#L782-L816
[2]: https://github.com/nomiclabs/hardhat/blob/22fef6a9b828b6f2432822e35484eb4f2b53dfdf/packages/hardhat-core/src/internal/hardhat-network/provider/node.ts#L1422-L1437

Reply:
@cgewecke awesome thanks for those links, i will take a look! i thought it would be good to introduce some kind of `stateAt(blockNumber): StateManager` functionality but didn't want to go that deep on this PR.